bin/run

```diff
@@ -81,33 +92,47 @@ with gfile.GFile(model_file, 'rb') as f:
         })
 
 inputs = []
-for i, spec in enumerate(input_spec):
-    def scope(i: int, spec: dict) -> Callable[[], tf.Tensor]:
-        def provide_input() -> tf.Tensor:
-            pl = tf.placeholder(tf.float32, shape=spec['shape'], name="api/{}".format(i))
-            return pl
-
-        return provide_input
-
-    inputs.append(scope(i, spec))
-
 
-if config.input_file != '':
-    # TODO: Currently this allows for only one input
-    # If we want to load models with multiple diffrent inputs
-    # We will need to save our inputs as a list of np.array
-    input_data = [np.load(config.input_file)]
-    assert len(input_data) == len(input_spec), 'the input_data length ({}) should be equal the number of inputs required by the loaded model ({})'.format(
-        len(input_data), len(input_spec))
+# If data_prep_path is set then extract data-preprocessing
+# graph from the SavedModel. Finally run it on the input-provider
+if config.data_prep_path:
+    data_file_names = list_files_from_dir(config.data_dir)
+    def provide_input():
+            return data_prep_from_saved_model(config.data_prep_path, 
```

Comment:
Running a session here (inside `data_pred_from_saved_model`) is still a bit weird to me. Not blocking, the semantics just bothers me. Is it the only way you could get device placement to work?

An alternative might be to load `gdef` outside of `provide_input` and pass that to `data_pred_from_saved_model` instead of `config.data_prep_path`. Maybe this won't result in the right placement though.

Reply:
I'd imagine as long as `tf.import_graph_def` happens within the device context when supplying this to `define_private_input`, that should be sufficient.  Otherwise, by passing the GraphDef directly we can iterate through it and manually override device NodeAttrs

Reply:
makes sense. Is it ok if we  run on `master` this to load `gdef`?

```
with tf.Session(graph=tf.Graph()) as sess:
        tf.saved_model.loader.load(sess, ['serve'],  config.data_prep_path)
    
gdef = sess.graph_def
```

then pass that to `data_pred_from_saved_model`. I have validated that the ops get placed correctly on the `input_provider`.

Reply:
Hm, perhaps we want to support both. Something like
```def data_prep_from_saved_model(saved_model_dir=None,  graph_def=None,...):
    if saved_model_dir is not None:
        # import on input provider
    if graph_def is not None:
        ...```

Reply:
@yanndupis if placement ends up being correct then I think it's alright to run that on master. Since you're loading into a new graph I wouldn't be surprised if the loaded savedmodel doesn't have any placements at all.

Reply:
Ok perfect, I will move forward with this. thanks.

Reply:
I was thinking that this preprocessing might be sensitive information, as we've considered in the past.  If we want to support that situation early, it would be best to support both loading the graph on master, and delegating it to another party so that it loads with the proper device placements (which seems to be the case based on the Tensorboard photos @yanndupis has shared).

Reply:
Most of the comments should be addressed. Now I am loading the saved_model with data prep in master however the data pre-processing is happening on the `input_device`. 

@jvmancuso, I think this a good point. It's just I don't know how to avoid having tf.Session in `data_prep_from_saved_model` if we want to delegate the load step to the `input_device`.  

btw, the protocol buffer file with the model architecture and the weights are loaded from the master. so it's a similar situation.

Reply:
fyi both of the two options we've seen here loads the preprocessing step on master; we currently only support in-graph distribution. this bothered me in the previous code because it hinted that loading might happen on a specific device. one way to remember this is that *all* python code currently executes on master, no exceptions.

Reply:
ah, that's unfortunate. getting the session outside of the provide_input fn makes the most sense